midend/flattenUnions.cpp

```diff
@@ -0,0 +1,359 @@
+                auto result = new IR::MethodCallExpression(right->srcInfo, IR::Type::Boolean::get(),
+                                                       isValid);
+                typeMap->setType(isValid, new IR::Type_Method(IR::Type::Boolean::get(),
+                                    new IR::ParameterList(), IR::Type_Header::isValid));
+
+                auto trueBlock = setInvalidforRest(a, lhs, huType, lhs->member.name, true);
+                auto method1 = new IR::Member(left->srcInfo, left,
+                                                  IR::ID(IR::Type_Header::setInvalid));
+                auto mc1 = new IR::MethodCallExpression(left->srcInfo, method1,
+                              new IR::Vector<IR::Argument>());
+                typeMap->setType(method1, new IR::Type_Method(IR::Type_Void::get(),
+                                    new IR::ParameterList(), IR::Type_Header::setInvalid));
+                auto ifFalse = new IR::MethodCallStatement(mc1->srcInfo, mc1);
+                auto ifStatement = new IR::IfStatement(a->srcInfo, result,
+                                           trueBlock->to<IR::BlockStatement>(), ifFalse);
+                return ifStatement;
+            }
+        } else if (auto leftMem = lhs->expr->to<IR::Member>()) {  // hdr.u1.h1 or u1.h1
+                // Handling hdr.u1.h1.data = <constant>
+            if (auto huType = leftMem->expr->type->to<IR::Type_HeaderUnion>()) {
+                // hdr.u1 or u1
+                if (leftMem->type->is<IR::Type_Header>() && right->is<IR::Constant>()) {
+                    // h1.data = constant
+                    return setInvalidforRest(a, leftMem, huType, leftMem->member.name, true);
+                }
+            }
+        }
+    }
+    return a;
+}
+
+const IR::Node* HandleValidityHeaderUnion::postorder(IR::IfStatement *a) {
+    if (auto mce = a->condition->to<IR::MethodCallExpression>()) {
+        return expandIsValid(a, mce);
+     }
+    return a;
+}
+
+const IR::Node* HandleValidityHeaderUnion::postorder(IR::SwitchStatement* a) {
+    if (auto mce = a->expression->to<IR::MethodCallExpression>()) {
+        return expandIsValid(a, mce);
+    }
+    return a;
+}
+
+// Assumes nested structs are flattened before this pass
+const IR::Node* HandleValidityHeaderUnion::postorder(IR::MethodCallStatement* mcs) {
+    auto mi = P4::MethodInstance::resolve(mcs->methodCall, refMap, typeMap);
+    if (auto a = mi->to<P4::BuiltInMethod>()) {
+        if (a->name == "setValid") {   // hdr.u.h1.setValid() or u.h1.setValid
+            if (auto m = a->appliedTo->to<IR::Member>()) {
+                if (auto huType = m->expr->type->to<IR::Type_HeaderUnion>())  // u.h1  or hdr.u.h1
+                    return setInvalidforRest(mcs, m, huType, m->member.name, false);
+            }
+        }
+    }
+    return mcs;
+}
+
+const IR::Node* HandleValidityHeaderUnion::postorder(IR::P4Action* action) {
```

Comment:
This pattern shows up a lot in the compiler, we should think whether it can be somehow automated...
But not in this PR.